include/sys/async_notify.h

```diff
@@ -0,0 +1,352 @@
+ * * a pointer to a specific client request structure, i.e. the one
+ *   that contains the async_notify structure.
+ * * the result of the operation, either as passed to
+ *   async_notify_finalize() or extracted afterwards using
+ *   async_notify_fetch_result().  Expected values are
+ *   service-specific, but the value shall be non-negative if the
+ *   operation succeeded, and negative if the operation failed.
+ */
+typedef void (*async_notify_generic_callback)();
+
+/**
+ * @brief State associated with notification for an asynchronous
+ * operation.
+ *
+ * Objects of this type are allocated by a client, which must use an
+ * initialization function (e.g. async_notify_init_signal()) to
+ * configure them.  Generally the structure is a member of a
+ * service-specific client structure, such as onoff_client.
+ *
+ * Control of the containing object transfers to the service provider
+ * when a pointer to the object is passed to a service function that
+ * is documented to take control of the object, such as
+ * onoff_service_request().  While the service provider controls the
+ * object the client must not change any object fields.  Control
+ * reverts to the client:
+ * * if the call to the service API returns an error;
+ * * when operation completion is posted (signalled or callback
+ *   invoked).  This may occur before the call to the service API
+ *   returns.
+ *
+ * After control has reverted to the client the notify object must be
+ * reinitialized for the next operation.
+ *
+ * The content of this structure is not public API to clients: all
+ * configuration and inspection should be done with functions like
+ * async_notify_init_callback() and async_notify_fetch_result().
+ * However, services that use this structure may access certain
+ * fields directly.
+ */
+struct async_notify {
+	union method {
+		/* Pointer to signal used to notify client.
+		 *
+		 * The signal value corresponds to the res parameter
+		 * of async_notify_callback.
+		 */
+		struct k_poll_signal *signal;
+
+		/* Generic callback function for callback notification. */
+		async_notify_generic_callback callback;
+	} method;
+
+	/*
+	 * Flags recording information about the operation.
+	 *
+	 * Bits below ASYNC_NOTIFY_EXTENSION_POS are initialized by
+	 * async notify API init functions like
+	 * async_notify_init_callback(), and must not by modified by
+	 * extensions or client code.
+	 *
+	 * Bits at and above ASYNC_NOTIFY_EXTENSION_POS are available
+	 * for use by service extensions while the containing object
+	 * is managed by the service.  They are not for client use,
+	 * are zeroed by the async notify API init functions, and will
+	 * be zeroed by async_notify_finalize().
+	 */
+	u32_t volatile flags;
+
+	/*
+	 * The result of the operation.
+	 *
+	 * This is the value that was (or would be) passed to the
+	 * async infrastructure.  This field is the sole record of
+	 * success or failure for spin-wait synchronous operations.
+	 */
+	int volatile result;
```

Comment:
> Also in case of callback notify using fetch result also does not make much sense since by the time it is fetched, async notify may already be used for something else and result is not valid.

Only if the user writes bad code.  Control of the object transfers to the user when the notification completes.  Nothing else is going to touch the content until the same object is given back to the service.

Note that this is not true of the signal result, becajuse the `k_poll_signal` object might be shared between multiple transition result objects, causing a wakeup when the first one completes.  This is another reason why having the result associated with the `async_notify` object rather than hidden away somewhere is useful.

And similarly for callback: it's passed to the callback as a convenience, but the callback may do nothing more than put the result on a queue for later processing.

> does it need to be `volatile`? It seems that when fetching result `flags` is checked and only when it indicates completion then result is read.

The `compiler_barrier()` I added recently in `async_notify_finalize()` shouldn't be necessary; if it's removed then `result` needs to be volatile to guarantee it's set after `flags`.  I'd like to remove that barrier.

Reply:
> Note that this is not true of the signal result, because the k_poll_signal object might be shared between multiple transition result objects

That sounds bad code to me to do that, as the one you mentioned earlier. How is the user going to know which objects the signal was about if these are sharing the same k_poll_signal?

Reply:
> Note that this is not true of the signal result, because the k_poll_signal object might be shared between multiple transition result objects, causing a wakeup when the first one completes.

If there are a shared signal being used and one task completes but async_notify client did not completed yet. How user can pend on that completion? Only by spining on fetching result?

Reply:
I was going to say "by using `async_notify_fetch_result` to see if the result is ready" but that contradicts my statement in https://github.com/zephyrproject-rtos/zephyr/pull/23229#discussion_r391558073 that the notification isn't complete until a callback is invoked.  So I'm going to have to change my position in that case.

Here, though, it's still valuable.  The other is that, because callbacks must be isr-ok and not sleep, there will be applications where the only thing the callback does is add the object that contains the notification object to a list for later processing in a thread.  In that case something still needs to store the specific result.

Reply:
yes, but then i don't think that it will be storing async_notifes rather another structure. Also in that case there can always be a dedicated structure holding async_notify and result. Or even use result field in if spinwait method was used. User can even utilize available space in `flags` to mark given async_notify object as the one holding result.

Reply:
It will probably be adding the object that contains the `async_notify`, as with the onoff manager.